DVR feature: Request and Response Hooks

jairmyree opened the pull request:
DVR now allows for hooks to be added to manipulate network requests and responses before they are persisted as interactions.

Commit:
Updates to DVR: now allows for hooks into the request and responses before they are persisted.

Commit:
Removed unnecessary(unimplemented) functions

Commit:
Minor edit to include Filter.swift

Commit:
Removed forced unwrapping from response data

Commit:
Missed the removal of this forced-unwrapping. Getting it now

tjprescott commented on DVR.xcodeproj/project.pbxproj:
```diff
@@ -18,6 +18,9 @@
 		2104F54D1DC658A60039CA14 /* URLResponse.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3647AFBF1B33602A00EF10D4 /* URLResponse.swift */; };
 		2104F54E1DC658A60039CA14 /* HTTPURLResponse.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3647AFC11B3363C400EF10D4 /* HTTPURLResponse.swift */; };
 		26D1E4181D1C30D500A20B76 /* SessionUploadTask.swift in Sources */ = {isa = PBXBuildFile; fileRef = B19D62641CB1860400E16D11 /* SessionUploadTask.swift */; };
+		3123728B26823CDE006C4D9E /* Filter.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3123728A26823CDE006C4D9E /* Filter.swift */; };
+		3123728C26823CDE006C4D9E /* Filter.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3123728A26823CDE006C4D9E /* Filter.swift */; };
+		3123728D26823CDE006C4D9E /* Filter.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3123728A26823CDE006C4D9E /* Filter.swift */; };
```
Curious that this appears 3 times?

jairmyree replied:
That's interesting. Especially since it's autogenerated. In Xcode I can't edit that file directly.

tjprescott commented on Sources/DVR/Filter.swift:
```diff
@@ -0,0 +1,50 @@
+//
+//  Filter.swift
+//  DVR
+//
+//  Created by Jáir Myree on 6/18/21.
+//  Copyright © 2021 Venmo. All rights reserved.
+//
```
This will need to change to use the Microsoft license header. If this is their stock header it's a little weird since they specify an MIT license: https://github.com/venmo/DVR/blob/master/LICENSE

jairmyree replied:
This is interesting, that was autogenerated so it is definitely weird that's how it came out. I'll replace it with the Microsoft header.

tjprescott replied:
It looks like the DVR source generally has no header, so I'd say stick to that and simply remove this, assuming we want to merge this back to Venmo eventually.

tjprescott commented on Sources/DVR/Filter.swift:
```diff
@@ -0,0 +1,50 @@
+    var beforeRecordResponse : ((inout Foundation.URLResponse, inout Data?) -> ())
+    var beforeRecordRequest : ((inout URLRequest) -> ())
+    
+    public init(headers: [String] = [], queryParameters: [String] = [], postDataParameters: [String] = [], replacer: String = "Redacted", requestHook : @escaping ((inout URLRequest) -> ()) = {_ in } , responseHook : @escaping ((inout Foundation.URLResponse, inout Data?) -> ()) = {_,_  in } ) {
```
Is `replacer` what is used in Python? To me it sounds like a function rather than a value. I would recommend `replacementString`. 

For the hooks, we should avoid using `inout` and made deliberate changes in AzureCore to avoid using it. It would be better to accept a URLRequest/Response and have the closure simply return a URLRequest/Response as well. This would most closely match the Python implementation.

It also would more closely match what we do in AzureCore regarding Pipeline stages:
https://github.com/Azure/azure-sdk-for-ios/blob/main/sdk/core/AzureCore/Source/Pipeline/PipelineStage.swift

jairmyree replied:
Sounds simple enough, I'll change it.

tjprescott commented on Sources/DVR/Filter.swift:
```diff
@@ -0,0 +1,50 @@
+        var adjustedQueryParameters : [String : String] = [:]
+        var adjustedPostDataParameters : [String : String] = [:]
+        
+        for header in headers {
+            adjustedHeaders[header] =  replacer
+        }
+        
+        for queryParam in queryParameters {
+            adjustedQueryParameters[queryParam] = replacer
+        }
+        
+        for postDataParam in postDataParameters {
+            adjustedPostDataParameters[postDataParam] = replacer
+        }
```
I'd recommend we add some unit tests for these if we are supporting the functionality.

tjprescott replied:
We would need to add unit tests anyhow to get this merged in the main repo, and you should get some practice writing unit tests :)

jairmyree replied:
I have the tests within my dummy project so I'll move them over.

tjprescott commented on Sources/DVR/Interaction.swift:
```diff
@@ -2,21 +2,24 @@ import Foundation
 struct Interaction {
 
-    // MARK: - Properties
+    // MARK: - Propertie6s
```
typo: Properties

tjprescott commented on Sources/DVR/Interaction.swift:
```diff
@@ -2,21 +2,24 @@ import Foundation
     // MARK: - Initializers
 
-    init(request: URLRequest, response: Foundation.URLResponse, responseData: Data? = nil, recordedAt: Date = Date()) {
+    init(request: URLRequest, response: Foundation.URLResponse, responseData: Data? = nil, recordedAt: Date = Date(), filter: Filter? = nil) {
```
It seems to me the filter should be associated with the session, not individual interactions.

tjprescott commented on Sources/DVR/Interaction.swift:
```diff
@@ -106,5 +109,8 @@ extension Interaction {
         self.recordedAt = Date(timeIntervalSince1970: recordedAt)
         self.responseData = Interaction.dencodeBody(response["body"], headers: response["headers"] as? [String: String])
+        self.filter = filter
+        self.filter?.beforeRecordRequest(&self.request)
+        self.filter?.beforeRecordResponse(&self.response,&self.responseData)
```
From my understanding of the framework, `beforeRecordRequest` should be called just before looking for an interaction match, while `beforeRecordResponse` should be called just before persisting a response. 

If an array of Interactions is initialized from a saved recording, everything should already be scrubbed, so you shouldn't need to run these methods at all in the initializer

jairmyree replied:
This is definitely something I want to talk about on our call.

tjprescott commented on Sources/DVR/Session.swift:
```diff
@@ -13,7 +13,9 @@ open class Session: URLSession {
     open var recordingEnabled = true
     open var recordMode : RecordingMode = .once
-
+    open var filter : Filter?
```
It might make things syntactically cleaner to specify a default do-nothing filter. From the initializer, that's what it seems you'd get anyhow.

jairmyree replied:
Yup, I made that change

tjprescott commented on Sources/DVR/SessionDataTask.swift:
```diff
@@ -11,7 +11,7 @@ final class SessionDataTask: URLSessionDataTask {
     // MARK: - Properties
 
-    weak var session: Session!
+    var session: Session!
```
Why the removal of `weak` here?

jairmyree replied:
So long story to be made short, I ran into an interesting error where depending on how long your hook was, the session would be de-initialized before the code finished running. I did a lot of research on how things are de-initialized and tracked it down to this assignment. To keep the session from getting de-initialized before the completion of the code, it needed a strong reference.

tjprescott commented on Sources/DVR/SessionDataTask.swift:
```diff
@@ -45,6 +45,10 @@ final class SessionDataTask: URLSessionDataTask {
 
     override func resume() {
         
+        guard session != nil else {
+            fatalError("This shouldn't happen")
+        }
+        
```
This isn't necessary, since if you dereference `session` and it's nil it will crash with a fatalError anyhow, but give a better error message.

tjprescott replied:
Arguably they shouldn't be using `!` anywhere, but this guard only makes sense if Session is optional

jairmyree replied:
This was necessary when testing for a weird error I encounter but it isn't anymore. I'll remove it

tjprescott commented on Sources/DVR/SessionDataTask.swift:
```diff
@@ -97,7 +101,7 @@ final class SessionDataTask: URLSessionDataTask {
             
             // Create interaction
-            this.interaction = Interaction(request: this.request, response: response, responseData: data)
+            this.interaction = Interaction(request: this.request, response: response, responseData: data, filter: this.session.filter)
```
My gut tells me we should not associate a filter with an interaction like this.

jairmyree replied:
I can explain my thought process better during our call.

tjprescott reviewed (changes requested):
This should be pretty easily unit testable. We should add simple unit tests. Some questions on implementation as well.

Commit:
Requested PR edits

Commit:
More requested PR edits

Commit:
Edits to PR

tjprescott commented on Tests/DVRTests/FilterTests.swift:
```diff
@@ -0,0 +1,32 @@
+//
+//  FilterTests.swift
+//  DVRTests-iOS
+//
+//  Created by Jáir Myree on 6/22/21.
+//  Copyright © 2021 Venmo. All rights reserved.
+//
+
```
Let's remove or replace this header as well

tjprescott commented on Tests/DVRTests/FilterTests.swift:
```diff
@@ -0,0 +1,32 @@
+
+    
+    
+    override func setUpWithError() throws {
+        // Put setup code here. This method is called before the invocation of each test method in the class.
+        
+    }
+
+    override func tearDownWithError() throws {
+        // Put teardown code here. This method is called after the invocation of each test method in the class.
+    }
+
+    func testExample() throws {
+        // This is an example of a functional test case.
+        // Use XCTAssert and related functions to verify your tests produce the correct results.
+    }
```
Assuming tests are forthcoming?

Commit:
PR Edits:
Removed convenience code (to be implemented at a later time). Created Unit Tests to test the ability to scrub using the filter functions

Commit:
Removed convenience initializer

tjprescott commented on Sources/DVR/Filter.swift:
```diff
@@ -31,18 +31,6 @@ public struct Filter {
-        self.beforeRecordResponse = responseHook
-    }
-    
     public init(replacements: [String:String] = [:], queryParameters: [String:String] = [:], postDataParameters: [String:String] = [:], requestHook : @escaping ((URLRequest) -> (URLRequest)) = {return $0} , responseHook : @escaping ((Foundation.URLResponse, Data?) -> (Foundation.URLResponse, Data?)) = {return ($0, $1) } ) {
```
Since we aren't implementing them right now, we can probably jettison replacements, queryParameters and postDataParameters here.
